### PR TITLE
Support gradient accumulation of fp16 in imperative mode

### PR DESCRIPTION
**Pull request by zhiqiu**
PR types: New features

PR changes: Others

Describe: Support gradient accumulation of fp16 in imperative mode

Note, original gradient accumulation of `float` and `double` uses `AXPY` of blas library, which does not support `float16`. So `float16` calculation uses `eigen` library.

### CODE REVIEW

**Review by chalsliu — DISMISSED** [review]

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 1]
```diff
@@ -100,6 +107,51 @@ class TensorAddFunctor : public boost::static_visitor<> {
...
+    }
+    case 6: {
+      operators::math::AddTensor<DeviceContext, T, 6> func;
+      func(dev_ctx, src, dst);
+      break;
+    }
+    default:
+      PADDLE_THROW(platform::errors::InvalidArgument(
+          "TensorAddImpl doesn't supports tensors whose ranks are greater "
+          "than 6."));
```
Comment (chenwhql) [inline paddle/fluid/imperative/gradient_accumulator.cc]: why doesn't support?

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 2]
```diff
@@ -133,8 +185,34 @@ void TensorAdd(const framework::Variable& src, framework::Variable* dst) {
...
+      TensorAddImpl<platform::CUDADeviceContext, platform::float16>(
+          src_tensor, dst_tensor, place);
+#else
+      PADDLE_THROW(platform::errors::PermissionDenied(
+          "Gradient accumulation of data type (%s) on place (%s) is not "
+          "supported in imperative mode",
+          framework::DataTypeToString(data_type), place));
+#endif
+    }
+    if (platform::is_cpu_place(place)) {
```
Comment (chenwhql) [inline paddle/fluid/imperative/gradient_accumulator.cc]: why not use `else if`?

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 3]
```diff
@@ -133,8 +185,34 @@ void TensorAdd(const framework::Variable& src, framework::Variable* dst) {
...
 
-  PADDLE_THROW("Not supported data type %s for AddTo",
-               framework::DataTypeToString(data_type));
+  if (data_type == framework::proto::VarType::FP16) {
+    if (platform::is_gpu_place(place)) {
+#ifdef PADDLE_WITH_CUDA
+      TensorAddImpl<platform::CUDADeviceContext, platform::float16>(
+          src_tensor, dst_tensor, place);
+#else
+      PADDLE_THROW(platform::errors::PermissionDenied(
```
Comment (chenwhql) [inline paddle/fluid/imperative/gradient_accumulator.cc]: may be `Unimplemented` is better than `PermissionDenied` here?

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/gradient_accumulator.cc]: Not sure, there are several operators which check `rank <= 6`, for example, `transpose` op. I will remove `switch...case...` and simplify the code.

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/gradient_accumulator.cc:200]: I will change that.

Comment (zhiqiu) [reply paddle/fluid/imperative/gradient_accumulator.cc]: I will change that, thx.

File: paddle/fluid/imperative/gradient_accumulator.cc (hunk 3 above)
Comment (zhiqiu) [reply paddle/fluid/imperative/gradient_accumulator.cc]: I will change that, thx.

File: paddle/fluid/imperative/gradient_accumulator.cc [hunk 4]
```diff
@@ -133,8 +150,26 @@ void TensorAdd(const framework::Variable& src, framework::Variable* dst) {
...
+#else
+      PADDLE_THROW(platform::errors::Unimplemented(
+          "Gradient accumulation of data type (%s) on place (%s) is not "
+          "supported in imperative mode",
+          framework::DataTypeToString(data_type), place));
+#endif
+    } else if (platform::is_cpu_place(place)) {
+      return TensorAddImpl<platform::CPUDeviceContext, platform::float16>(
+          src_tensor, dst_tensor, place);
+    }
```
Comment (chenwhql) [inline paddle/fluid/imperative/gradient_accumulator.cc:167]: why remove `cuda_pinned_place` check? do we need an `else` branch here?

**Review by chenwhql — APPROVED** [review]
LGTM

**Pull request merged**